src/Lean/Elab/Tactic/BVAckermannize.lean

```diff
@@ -0,0 +1,407 @@
+  return (fn, args)
+end Function
+/--
+NOTE: is it sensible to hash an array of arguments?
+We may want to use something like a trie to index these.
+Consider swiching to something like `Trie`.
+-/
+abbrev ArgumentList := Array Argument
+/--
+The data stored for an ackermannized call to allow us to build proofs.
+-/
+structure CallVal where
+  /-- The free variable `ack_fx₁...xₙ := (f x₁ x₂ ... xₙ)`. -/
+  fvar : FVarId
+  /-- heqProof : The proof that `ack_fx₁...fxₙ = f x₁ x₂ ... xₙ` (name/fvar = value/expr). -/
+  heqProof : Expr
+deriving Hashable, BEq, Inhabited
+
+namespace CallVal
+
+instance : ToMessageData CallVal where
+  toMessageData cv := m!"{Expr.fvar cv.fvar} ({cv.heqProof})"
+
+end CallVal
+structure State where
+  /--
+  A mapping from a function `f`, to a map of arguments `x₁ ... xₙ`, to the information stored about the call.
+  This is used to generate equations of the form `x₁ = y₁ → x₂ = y₂ → ... → xₙ = yₙ → ack_fx₁...xₙ = ack_fy₁...yₙ on-demand.
+  -/
+  fn2args2call : Std.HashMap Function (Std.HashMap ArgumentList CallVal) := {}
+  /-- A counter for generating fresh names. -/
+  gensymCounter : Nat := 0
+
+def State.init : State where
+
+abbrev AckM := StateRefT State MetaM
+
+namespace AckM
+
+def run (m : AckM α) : MetaM α := m.run' State.init
+
+/-- Generate a fresh name. -/
+def gensym : AckM Name := do
+  modify fun s => { s with gensymCounter := s.gensymCounter + 1 }
+  return Name.mkNum (Name.mkSimple "ackConst") (← get).gensymCounter
+
+def withContext (g : MVarId) (ma : AckM α) : AckM α := g.withContext ma
+
+/-- Get the calls to a function `fn`. -/
+def getCallMap (fn : Function) : AckM (Std.HashMap ArgumentList CallVal) := do
+  return (← get).fn2args2call.getD fn {}
+
+/-- Get the calls to a function `fn`. -/
+def getCallVal? (fn : Function) (args : Array Argument) : AckM (Option CallVal) := do
+  let calls ← getCallMap fn
+  if let .some val := calls.get? args then
+    return some val
+  return none
+
+structure IntroDefResult where
+  -- the new name/fvar of the defn.
+  defn : FVarId
+  -- a proof 'hdefn : name = value
+  eqProof : FVarId
+
+/-
+Introduce a new definition with name `name : hdefTy` into the local context,
+and return the FVarId of the new definition in the new goal (the MVarId) returned.
+-/
+def introDefExt (g : MVarId) (name : Name) (hdefTy : Expr) (hdefVal : Expr) : AckM (IntroDefResult × MVarId) := do
+  withContext g do
+    let g ← g.assertExt name (hName := Name.str name "value") hdefTy hdefVal
+    let (defn, g) ← g.intro1P
+    let (eqProof, g) ← g.intro1P
+    return ({ defn, eqProof}, g)
+
+/-- Insert the CallVal `cv` at `(fn, args)` into the state. -/
+private def _insertCallVal (fn : Function) (args : ArgumentList) (cv : CallVal) : AckM Unit := do
+  let calls ← getCallMap fn
+  modify fun s => { s with fn2args2call := s.fn2args2call.insert fn (calls.insert args cv) }
+
+/--
+Replace a call to the function `f` with an `fvar`. Since the `fvar` is defeq to the call,
```

Comment:
It is not clear without looking at the implementation of the function that the `fvar` you are referring to here is the returned `CallVal`. The remark about having to add another hypothesis later is also not clear to me as `CallVal` already contains a proof of this, why don't we add it on the spot when the `CallVal` got created and store the `FVarId` in `CallVal`? After all the `CallVal` seems to be cached anyway so there should not be a risk of duplicating these theorems?

Reply:
I rephrased this, and explain that we need to later add the *ackermannization* theorems for the given call, not the equation theorem. Thanks for the critique.